packages/headless/src/app/commerce-engine/commerce-engine.ssr.ts

```diff
@@ -0,0 +1,221 @@
+/**
+ * Utility functions to be used for Commerce Server Side Rendering.
+ */
+import {UnknownAction} from '@reduxjs/toolkit';
+import {stateKey} from '../../app/state-key';
+import {buildProductListing} from '../../controllers/commerce/product-listing/headless-product-listing';
+import {buildSearch} from '../../controllers/commerce/search/headless-search';
+import type {Controller} from '../../controllers/controller/headless-controller';
+import {LegacySearchAction} from '../../features/analytics/analytics-utils';
+import {createWaitForActionMiddleware} from '../../utils/utils';
+import {
+  buildControllerDefinitions,
+  composeFunction,
+  createStaticState,
+} from '../ssr-engine/common';
+import {
+  ControllerDefinitionsMap,
+  InferControllerPropsMapFromDefinitions,
+} from '../ssr-engine/types/common';
+import {
+  EngineDefinition,
+  EngineDefinitionOptions,
+} from '../ssr-engine/types/core-engine';
+import {
+  CommerceEngine,
+  CommerceEngineOptions,
+  buildCommerceEngine,
+} from './commerce-engine';
+
+export interface SSRCommerceEngineOptions extends CommerceEngineOptions {
+  /**
+   * The type of the solution the engine will be used for
+   * - 'search': Indicates that the engine is used for Search.
+   * - 'product-listing': Indicates that the engine is used for Product listing.
+   */
+  solutionType: 'search' | 'listing';
+}
+
+/**
+ * The SSR commerce engine.
+ */
+export interface SSRCommerceEngine extends CommerceEngine {
+  /**
+   * Waits for the search to be completed and returns a promise that resolves to a `SearchCompletedAction`.
+   */
+  waitForSearchCompletedAction(): Promise<SearchCompletedAction>;
```

Comment:
To better reflect that this works with listings, where we don't execute searches, I would propose we rename this
```suggestion
  waitForRequestCompletedAction(): Promise<SearchCompletedAction>;
```